### ISSUE TITLE
HeapPtr<T> Portability

### ISSUE THREAD

**Issue by clemenswasser** [luajit-remake/luajit-remake#8]
Hello,

I am genuinely fascinated by this project and have been working on enhancing the CMake build system to clean things up and support Windows. One problem I am currently facing is the missing portability of the `HeapPtr<T>` from `heap_ptr_utils.h`. This issue not only affects Windows, but also poses challenges when using compilers other than Clang (e.g., GCC) due to the absence of `__attribute__(address_space(...))`. Given its impact on portability beyond just Windows, I believe that having an open discussion would be beneficial.

To address this portability issue, I thought of some possible solutions:
 - MSVC has the intrinsic functions [\_\_readgs*](https://learn.microsoft.com/en-us/cpp/intrinsics/readgsbyte-readgsdword-readgsqword-readgsword?view=msvc-160), [\_\_writegs*](https://learn.microsoft.com/en-us/cpp/intrinsics/writegsbyte-writegsdword-writegsqword-writegsword?view=msvc-160) which could be a possible solution, at least for Windows. This would however involve changing the `HeapPtr<T>` to some sort of pointer wrapper class, which is incompatible with the current design and would require many changes across the codebase. I also did some tests with the intrinsic functions, but they don't seem to work (probably because the GS register is used for thread local storage on 64-bit Windows).
 -  A solution compatible with all compilers could involve using the [\_\_readgsbase_*](https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html#text=_readgsbase_&ig_expand=5353,5352,5355) intrinsic functions with additional pointer arithmetic. This would however also need a change to the `HeapPtr<T>` definition, with the difficulties mentioned above. Also, I don't know how much overhead this would introduce, as a relative GS read is probably faster than reading the GS Base doing arithmetic and then reading from the final address.
 - Since the GS register only exists on x86, we would need an entirely different mechanism for other CPU architectures anyway. Hence, we could potentially emulate the GS register using thread local storage of a base address on other architectures.

I came across an informative article regarding the portable use of the GS register, which you might find valuable: https://mary.rs/lab/gsbase/

I also compared the performance of the possible solutions, described above [here](https://github.com/clemenswasser/gs_register_bench).
At least on Linux, an implementation that stores the current base address in thread local storage appears to be the most efficient option. It achieves nearly identical performance to reading directly from a buffer, with the buffer base address residing in a register.
The direct gs offset read seems to be 2.26x **slower** (this is the current non-portable approach) and reading the gs base address and then computing and dereferencing the pointer is 6.3x **slower**.
I sadly can't test the ones using the GS register on Windows, since that doesn't work.

Best regards,
Clemens

**Comment by sillycross** [comment]
Hi Clemens,

Thanks for posting the issue! The investigation you did is impressive: I was totally not anticipating that someone would experiment with the code in detail and post a good issue like this :)

You are correct, the GS-register boxing scheme is a big design mistake that introduces tons of portability issues (thought it might indeed bought us a few percents of performance).

However, that design was implemented at the very beginning, so at this moment it will require a lot of efforts to clean it up. If you are interested in doing this I'm happy to provide help, but keep in mind that it's not going to be a simple fix, and will take a lot of devotion.

I'll first give you an overview on the current design (and thus the work needed to improve it):

**VM Heap Layout and GS register**

Currently the whole VM heap is laid out in a flat 18GB address range (see [code here](https://github.com/luajit-remake/luajit-remake/blob/63415c01dafad6be1ded756e99432a5151dba044/runtime/vm.h#L502), and a simplified illustration below), with the GS register pointing to the 'base': a convenient place that allows us to access everything easily. 
```
[ 12 GB User Heap ] [ 4GB padding ] [ class VM ] [ 2 GB System Heap ]
                                    ^ GS
```
The whole address range is reserved at once (but not backed by physical memory, and is not readable/writable), and then we use mmap's MAP_FIXED option to actually map physical memory pages onto the address range as needed.

The GS register trick:

1. Allows us to represent a system-heap pointer in 4 bytes. This is relatively important: it not only saves space, but also allows inline caching to compare the pointer values easily. With 4 bytes the comparison can be done by `cmpl $1234567, %eax` but with 8 bytes you need to do `movabsq $XXXXXXXXXXXX, %rbx; cmpq %rbx, %rax` which is much longer and slower.

2. Allows us to directly access the heap location from the boxed value using GS-related addressing mode: say `%rax` holds a boxed value, if it represents a heap entity, then the address of the heap entity is just `gs:%rax`. This allows most of the heap access to be done directly in 1 instruction using x64 addressing mode. 

3. Allows us to not pass the VM pointer everywhere. Slow paths can just get GS by reading `gs:0x0` (the VM class always sits at `GS:0x0`, and offset 0 of the VM class always stores the GS pointer value itself, so reading `gs:0x0` is allows you to get the GS pointer value). 

As you can see, the trick is indeed helpful performance-wise, but it also locks us to x86-64 Linux, as you have noticed.

The root fix is to redesign the scheme so that the VM is explicitly passes as in a CPU register in the interpreter/JIT, and then change the HeapPtr dereference logic do to `vm_base + value` instead of `gs:value`. However, as you can see, this will require a lot of changes as `HeapPtr` is currently used in many places (unfortunately).

The VM pointer still should not be explicitly passed to everywhere in the slow path (as that could be a sweeping change). The solution is to store the VM pointer in the thread-local storage, i.e., `VM::GetActiveVMForCurrentThread()` will not read `gs:0x0`, but just access the `thread_local VM* t_vm`.

Regarding your questions:

> ... but also poses challenges when using compilers other than Clang (e.g., GCC)

Due to fundamental design of Deegen (which uses Clang to compile C++ files to LLVM IR, then do various transformations on LLVM IR level to generate the interpreter/JIT), supporting non-Clang compiler is already out of the question. 

It might still beneficial to expose a C++ API header (which can be used by non-Clang compilers) to allow LJR to be embedded as a library, but this requires a ton of design and work (e.g., we don't even have implemented the Lua C API yet, and efficiently support FFI is an active research topic on its own, for example, [Cinder](https://github.com/facebookincubator/cinder/) has been focusing on doing this for years). So I wouldn't think too much about this right now.

> MSVC has the intrinsic functions [__readgs*](https://learn.microsoft.com/en-us/cpp/intrinsics/readgsbyte-readgsdword-readgsqword-readgsword?view=msvc-160), [__writegs*](https://learn.microsoft.com/en-us/cpp/intrinsics/writegsbyte-writegsdword-writegsqword-writegsword?view=msvc-160) which could be a possible solution...

Have you tried to use FS segmentation register on Windows? I think FS is available for use by user applications, but unfortunately I haven't used Windows for 10+ years (so my familiarity to Windows remains in the Windows XP days...), so I have no idea if my knowledge is correct or not, and can't anticipate what challenges you would encounter. 

> Since the GS register only exists on x86, we would need an entirely different mechanism for other CPU architectures anyway.

This is correct. To support another architecture like ARM, the "root fix" I described above must be applied.

> The direct gs offset read seems to be 2.26x slower (this is the current non-portable approach) 

You definitely messed up with something here. TLS-read is a FS-prefixed read, there is no reason that a FS-prefixed read is 2.26x faster than a GS-prefixed read. You need to look at the assembly to know what the compiler did.

To summarize, I can suggest two directions:
1. If your focus is to make LJR work for Windows, you should test if the FS segmentation register works. If it works then it's good.
2. If (1) doesn't work out, or if you want to make some non-trivial changes to LJR :smiley: , you can consider implementing the "root fix" to make the boxing scheme no longer dependent on GS on non-Linux environments. That would be an important improvement, but it's not going to be simple: you will need to dig deep into Deegen in many places. Of course I'm happy to help, but you have to do the work, which is likely a lot :)

**Comment by zNoctum** [comment]
Hi so i wanted to ask if all of the different HeapPtrs i.e. SystemHeapPtr, UserHeapPtr, ... can all be folded into one new HeapPtr class?

**Comment by zNoctum** [comment]
So i have started adding the ability to pass the VM base pointer in register but, now have encountered a weird problem, where even though no refrence to the `thread_local` that i use for the vm pointer shows up in any of the `audit*/` dirs it still encounters TLS relocation records in the baseline JIT stencil creation and thus fails. It would be great if you could take a look at my [fork](https://github.com/zNoctum/luajit-remake) and could help me to narrow the source of this problem down.

**Comment by sillycross** [comment]
What if you just use a normal global instead of thread_local as a temporary workaround? Issues like this are only worth spending time if you can make the primary goal work...

**Comment by zNoctum** [comment]
A normal global works but i don't see the benefit because eventually the base pointer needs to be `thread_local` to achieve feature parity with the current state of things or do you mean that I should use a normal global as a fix until it works decently enough to focus on that?

**Comment by sillycross** [comment]
> do you mean that I should use a normal global as a fix until it works decently enough to focus on that?

Yes exactly.

**Comment by zNoctum** [comment]
I have done some testing and the new implementation is 8% slower than using the segmented memory. Do you have a idea where I could make up a bit of that performance @sillycross?
![image](https://github.com/user-attachments/assets/15be4d5e-d24e-42d9-a679-dac9e89d0b89)
